File: Sorting/Quick_Sort/quicksort.py
import math
import os

def partition(array,left,right,pivotindex):
    pivot = array[pivotindex]
    # print("\n {pivot}").format(pivot=pivot)
    # print("\n Array before {0}").format(array)
    array[right],array[pivotindex] = array[pivotindex],array[right]
    # print("\n Array after {0}").format(array)

    storeindex=left
    
    for i in xrange(left,right):
        if(array[i]<array[right]):
            array[i],array[storeindex] = array[storeindex],array[i]
            storeindex = storeindex + 1
    array[storeindex],array[right] = array[right],array[storeindex]
    return storeindex

def quicksort(array,left,right):
    if left<right:
        mid = (left+right)/2
        newpivotindex = partition(array,left,right,mid)
        print array
        quicksort(array,left,newpivotindex-1)
        quicksort(array,newpivotindex+1,right)
    return array

def main():
    array = [10,9,8,7,6,5,4,8,3,2]
    print (quicksort(array,0,len(array)-1))

if __name__ == '__main__':
    main()

File: Sorting/Quick_Sort/QuickSort.java
package sorting;

import java.util.ArrayList;

public class QuickSort {

	//ArrayList<Integer> array = new ArrayList<Integer>();
	Integer[] arrayToBeSorted = new Integer[10];
	
	public QuickSort(Integer[] arrayToBeSorted){
		//this.array = array;
		this.arrayToBeSorted = arrayToBeSorted;
	}
	
	public void swap(int a, int b){
		int temp = arrayToBeSorted[a];
		arrayToBeSorted[a] = arrayToBeSorted[b];
		arrayToBeSorted[b] = temp;	
	}
	
	public int partition(int left, int right, int pivotIndex){
		int pivot = arrayToBeSorted[pivotIndex];
		swap(pivotIndex, right);
		int storeIndex = left;
		for(int i=left; i<arrayToBeSorted.length -1; i++){
			if(arrayToBeSorted[i] < arrayToBeSorted[right]){
				swap(i, storeIndex);
				storeIndex++ ;
			}
		}
		swap(storeIndex, right);
		return storeIndex;
	}
	
	public Integer[] recursiveQS(int left,int right, QuickSort qs){
		
		if(left < right){
			int pivotIndex = (left+right)/2;
			int partitionIndex = qs.partition(left, right, pivotIndex);
			System.out.println("Partition Index : " + partitionIndex);
			qs.recursiveQS(left, partitionIndex-1, qs);
			qs.recursiveQS(partitionIndex+1, right, qs);
		}
		return arrayToBeSorted;
	}
	
	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Integer[] arrayToBeSorted = {9,8,7,6,5,4,3,2,1,0};
		QuickSort qs = new QuickSort(arrayToBeSorted);
		//System.out.println(qs.partition(0, arrayToBeSorted.length-1, 9));
		Integer[] sortedArray = qs.recursiveQS(0, arrayToBeSorted.length-1, qs);
		for(int i =0; i<arrayToBeSorted.length; i++){
			System.out.print(" "+sortedArray[i]+" ");
		}
	}

}
